libvisual/libvisual/lv_video.cpp

```diff
@@ -292,6 +293,110 @@ namespace LV {
       return true;
   }
 
+  #ifdef LV_ENABLE_TESTS
+  bool Video::validate() const
+  {
+      // Validate dimensions.
+
+      if (m_impl->width <= 0) {
+          visual_log (VISUAL_LOG_ERROR, "Video has a zero or negative width (%d).", m_impl->width);
+          return false;
+      }
+
+      if (m_impl->height <= 0) {
+          visual_log (VISUAL_LOG_ERROR, "Video has a zero of negative height (%d).", m_impl->height);
+          return false;
+      }
+
+      if (m_impl->pitch <= 0) {
+          visual_log (VISUAL_LOG_ERROR, "Video has a zero of negative pitch (%d).", m_impl->pitch);
+          return false;
+      }
+
+      // Validate extents.
+
+      if (m_impl->width != m_impl->extents.width) {
+          visual_log (VISUAL_LOG_ERROR, "Width (%d) and width of extents (%d) do not agree.",
+                      m_impl->width, m_impl->extents.width);
+          return false;
+      }
+
+      if (m_impl->height != m_impl->extents.height) {
+          visual_log (VISUAL_LOG_ERROR, "Height (%d) and height of extents (%d) do not agree.",
+                      m_impl->height, m_impl->extents.height);
+          return false;
+      }
+
+      // Validate parent-child relations.
+
+      if (m_impl->parent) {
+          auto parent_impl = m_impl->parent->m_impl.get ();
+          if (!parent_impl->extents.contains (m_impl->extents)) {
+              visual_log (VISUAL_LOG_ERROR, "Sub-video is not fully contained by parent.");
+              return false;
+          }
+      }
+
+      if (m_impl->depth != VISUAL_VIDEO_DEPTH_GL) {
+          // Validate depth and bytes per pixel consisteency.
+
+          if (m_impl->bpp != visual_video_depth_bpp (m_impl->depth) / 8) {
+              visual_log (VISUAL_LOG_ERROR, "Depth (%s) is not consistent with BPP (%d).",
+                          visual_video_depth_name (m_impl->depth), m_impl->bpp);
+              return false;
+          }
+
+          // Validate pixel row table
+
+          auto pixel_row_ptr = static_cast<uint8_t const*> (get_pixels ());
+
+          for (int y = 0; y < m_impl->height; y++) {
+              if (m_impl->pixel_rows[y] != pixel_row_ptr) {
+                  visual_log (VISUAL_LOG_ERROR, "Pixel row pointer table is wrong at y=%d.", y);
+                  return false;
+              }
+
+              pixel_row_ptr += m_impl->pitch;
```

Comment:
I'd inline this to `expected_pixel_row_ptr = pixel_row_ptr + y * m_impl->pitch` and compare against that. The current code only makes full sense once the reader arrives at line 359 which is arguably later than ideal.

Reply:
I agree, a direct calculation is better for validation.